optics-th/src/Optics/TH/Internal/Product.hs

```diff
@@ -95,9 +96,26 @@ makeFieldOpticsForDatatype rules info =
   normFieldLabels :: Traversal [(Name,[(a,Type)])] [(Name,[(b,Type)])] a b
   normFieldLabels = traversed % _2 % traversed % _1
 
-  -- Map a (possibly missing) field's name to zero-to-many optic definitions
-  expandName :: [Name] -> Maybe Name -> [DefName]
-  expandName allFields = concatMap (_fieldToDef rules tyName allFields) . maybeToList
+-- | Map a (possibly missing) field's name to zero-to-many optic definitions
+expandName :: LensRules -> Name -> [D.ConstructorInfo] -> [Name] -> Maybe Name -> [DefName]
+expandName rules tyName cons allFields =
+    concatMap (_fieldToDef rules tyName allFields . stripSel) . maybeToList
+  where
+    -- When DuplicateRecordFields is enabled, reified datatypes contain
+    -- "mangled" field names that look like $sel:foo:MkT where foo is the field
+    -- name and MkT is the first data constructor.  We strip off the prefix and
```

Comment:
Is `MkT` always the first constructor, even in cases like `data T = MkT1 { name1 :: String } | MkT2 { name2 :: String }` `name2` will be mangled as `$sel:name2:mkT1`?

Reply:
Yes, it's just the first constructor in the type, regardless of whether the field occurs or not. I've tweaked the test to check this, and cleaned up the code a bit. In particular it now checks that the suffix is the expected one, which should lead to easier to understand errors if GHC's behaviour should ever change.